scipy/cluster/_hierarchy.pyx

```diff
@@ -754,6 +754,153 @@ def linkage(double[:] dists, np.npy_int64 n, int method):
                 D[condensed_index(n, i, x)] = NPY_INFINITYF
     return Z_arr
 
+
+cdef find_min_dist(int n, double[:] D, int[:] size, int x):
+    cdef double current_min = NPY_INFINITYF
+    cdef int y = -1
+    cdef int i
+    cdef double dist
+
+    for i in range(x + 1, n):
+        if size[i] == 0:
+            continue
+
+        dist = D[condensed_index(n, x, i)]
+        if dist < current_min:
+            current_min = dist
+            y = i
+
+    return y, current_min
```

Comment:
It's kind of awkward to be honest to split this method.

I don't know, maybe create a simple `pair` C struct?